lib/Db/RowMapper.php

```diff
@@ -133,295 +327,340 @@ private function resolveSearchValue(string $unresolvedSearchValue, string $userI
+	 * @param mixed $value the value should be parsed to the correct format within the row service
+	 * @param Column $column
+	 * @throws InternalError
+	 */
+	private function updateCell(RowCellSuper $cell, RowCellMapperSuper $mapper, $value, Column $column): void {
+		$v = $this->formatValue($column, $value, 'in');
+		$cell->setValueWrapper($v);
+		$this->updateMetaData($cell);
+		$mapper->updateWrapper($cell);
 	}
 
-	private function getAllColumnIdsFromView(View $view, IQueryBuilder $qb): array {
-		$neededColumnIds = [];
-		$filters = $view->getFilterArray();
-		$sorts = $view->getSortArray();
-		foreach ($filters as $filterGroup) {
-			foreach ($filterGroup as $filter) {
-				$neededColumnIds[] = $filter['columnId'];
-			}
-		}
-		foreach ($sorts as $sortRule) {
-			$neededColumnIds[] = $sortRule['columnId'];
+	/**
+	 * @throws InternalError
+	 */
+	private function insertOrUpdateCell(int $rowId, int $columnId, $value): void {
+		$cellMapperClassName = 'OCA\Tables\Db\RowCell'.ucfirst($this->columns[$columnId]->getType()).'Mapper';
+		/** @var RowCellMapperSuper $cellMapper */
+		try {
+			$cellMapper = Server::get($cellMapperClassName);
+		} catch (NotFoundExceptionInterface|ContainerExceptionInterface $e) {
+			$this->logger->error($e->getMessage(), ['exception' => $e]);
+			throw new InternalError(get_class($this) . ' - ' . __FUNCTION__ . ': '.$e->getMessage());
 		}
-		$neededColumnIds = array_unique($neededColumnIds);
-		if ($this->platform === IColumnTypeQB::DB_PLATFORM_PGSQL) {
-			foreach ($neededColumnIds as $columnId) {
-				if ($columnId >= 0) {
-					/** @psalm-suppress ImplicitToStringCast */
-					$qb->leftJoin("t1", $qb->createFunction('json_array_elements(t1.data)'), 'c' . intval($columnId), $qb->createFunction("CAST(c".intval($columnId).".value->>'columnId' AS int) = ".$columnId));
-					// TODO Security
-				}
-			}
+		try {
+			$cell = $cellMapper->findByRowAndColumn($rowId, $columnId);
+			$this->updateCell($cell, $cellMapper, $value, $this->columns[$columnId]);
+		} catch (DoesNotExistException $e) {
+			$this->insertCell($rowId, $columnId, $value);
+		} catch (MultipleObjectsReturnedException|Exception $e) {
+			$this->logger->error($e->getMessage(), ['exception' => $e]);
+			throw new InternalError(get_class($this) . ' - ' . __FUNCTION__ . ': '.$e->getMessage());
 		}
-		return $neededColumnIds;
 	}
 
 	/**
-	 * @throws MultipleObjectsReturnedException
-	 * @throws DoesNotExistException
-	 * @throws Exception
+	 * @param Column[] $columns
 	 */
-	public function findNext(int $offsetId = -1): Row {
-		$qb = $this->db->getQueryBuilder();
-		$qb->select('t1.*')
-			->from($this->table, 't1')
-			->where($qb->expr()->gt('id', $qb->createNamedParameter($offsetId)))
-			->setMaxResults(1)
-			->orderBy('id', 'ASC');
-
-		return $this->findEntity($qb);
+	private function setColumns(array $columns): void {
+		foreach ($columns as $column) {
+			$this->columns[$column->getId()] = $column;
+		}
```

Comment:
Is it safe that this instance is having state? It is instantiated once via DI. Could this interfere with another earlier or later call?